src/UserSpaceInstrumentation/AllocateInTracee.h

```diff
@@ -9,30 +9,92 @@
-// error is returned.
-// If `address` is zero the placement of memory will be arbitray (compare the documenation of mmap:
-// https://man7.org/linux/man-pages/man2/mmap.2.html). Assumes we are already attached to the tracee
-// `pid` e.g. using `AttachAndStopProcess`.
-[[nodiscard]] ErrorMessageOr<uint64_t> AllocateInTracee(pid_t pid, uint64_t address, uint64_t size);
-
-// Free address range previously allocated with AllocateInTracee using munmap.
-// Assumes we are already attached to the tracee `pid` e.g. using `AttachAndStopProcess`.
-[[nodiscard]] ErrorMessageOr<void> FreeInTracee(pid_t pid, uint64_t address, uint64_t size);
-
-// Same as `AllocateInTracee` above but wraps the memory in a unique_resource that handles
-// deallocating the memory. Note that we still need to be attached (or attached again) to the tracee
-// when the unique_resource goes out of scope.
-[[nodiscard]] ErrorMessageOr<orbit_base::unique_resource<uint64_t, std::function<void(uint64_t)>>>
-AllocateInTraceeAsUniqueResource(pid_t pid, uint64_t address, uint64_t size);
+// Represents a chunk of memory in the tracee.
+// The only way to instantiate the class is via the factory function `Create`. We move around
+// MemoryInTracee exclusively in unique_ptr. This way we can store instances inside stl containers.
+// The memory gets allocated in writable state. In case one wants to execute code in the segment, it
+// needs to be made executable with `EnsureMemoryExecutable` later. `Free` deallocates the memory.
+// This needs to be done manually - if MemoryInTracee goes out of scope without being freed the
+// memory inthe tracee will leak.
+// Note that for each of the non-const member functions we need to
+// execute code in the tracee. So we need to be attached; the tracee needs to be stopped.
+class MemoryInTracee {
+ public:
+  // We move around MemoryInTracee exclusively in unique_ptr so we don't need it to be copyable or
+  // movable. The only way to create a MemoryInTracee is to use the factory function below which
+  // uses the private constructor.
+  MemoryInTracee() = delete;
+  MemoryInTracee(const MemoryInTracee&) = delete;
+  MemoryInTracee(MemoryInTracee&&) = delete;
+  MemoryInTracee& operator=(const MemoryInTracee&) = delete;
+  MemoryInTracee& operator=(MemoryInTracee&&) = delete;
+
+  // Allocate `size` bytes of memory in the tracee's address space using mmap. The memory will have
+  // write permissions. The memory allocated will start at `address`. `address` needs to be aligned
+  // to page boundaries. If the memory mapping can not be placed at `address` an error is returned.
+  // If `address` is zero the placement of memory will be arbitrary (compare the documentation of
+  // mmap: https://man7.org/linux/man-pages/man2/mmap.2.html). Assumes we are already attached to
+  // the tracee `pid` using `AttachAndStopProcess`.
+  [[nodiscard]] static ErrorMessageOr<std::unique_ptr<MemoryInTracee>> Create(pid_t pid,
```

Comment:
Also this doesn't have to be a unique_ptr or does it?

Reply:
I know based on discussion that you might want `MemoryInTracee` to be a `unique_ptr` to store in containers, but in the majority of cases (all the cases in this change) this doesn't apply. The way I see it this type should be moveable and stack-allocatable, and then you can have two different factory methods to return you a `MemoryInTracee` directly, or a `std::unique_ptr<MemoryInTracee>` when you specifically need it.

Reply:
Right, the unique_ptr is there in order to be able to store these things in vectors.

I could additionally provide a factory returning an object on the stack; there is no technical reason not to make `MemoryInTracee` movable. I would totally do that if `MemoryInTracee` was a lightweight, frequently-allocated class. Then the performance advantage would matter. But since every construction requires me to invoke ptrace / waitpid multiple times this doesn't matter at all. Also I think it is nicer to only have one factory instead of two. And I use this in about five places. I just don't need more complications.

Reply:
> there is no technical reason

No technical reason and no performance reason, it was about more idiomatic C++. But of course what you have works too, in the end this is fine for me. I agree it's nicer to only have one factory. Again, maybe @dimitry- also has an opinion given his first comment at the top of this thread?